staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go

```diff
@@ -330,54 +318,73 @@ func isIncompatibleServerError(err error) bool {
+// as a slice of pointer to resource.Info. The resource.Info contains the object
+// and some other denormalized data. This function should not be called until
+// AFTER the "complete" and "validate" methods have been called to ensure that
+// the ApplyOptions is filled in and valid. Returns an error if the resource
+// builder returns an error retrieving the objects.
+func (o *ApplyOptions) GetObjects() ([]*resource.Info, error) {
+	if !o.objectsCached {
+		// include the uninitialized objects by default if --prune is true
+		// unless explicitly set --include-uninitialized=false
+		r := o.Builder.
+			Unstructured().
+			Schema(o.Validator).
+			ContinueOnError().
+			NamespaceParam(o.Namespace).DefaultNamespace().
+			FilenameParam(o.EnforceNamespace, &o.DeleteOptions.FilenameOptions).
+			LabelSelectorParam(o.Selector).
+			Flatten().
+			Do()
+		if err := r.Err(); err != nil {
+			return nil, err
+		}
+		infos, err := r.Infos()
+		if err != nil {
+			return nil, err
+		}
+		o.objects = infos
+		o.objectsCached = true
+	}
+	return o.objects, nil
+}
+
+// SetObjects stores the set of objects (as resource.Info) to be
+// subsequently applied.
+func (o *ApplyOptions) SetObjects(infos []*resource.Info) {
+	o.objects = infos
+	o.objectsCached = true
+}
+
 // Run executes the `apply` command.
 func (o *ApplyOptions) Run() error {
-	var openapiSchema openapi.Resources
-	if o.OpenAPIPatch {
-		openapiSchema = o.OpenAPISchema
-	}
 
 	dryRunVerifier := &DryRunVerifier{
 		Finder:        cmdutil.NewCRDFinder(cmdutil.CRDFromDynamic(o.DynamicClient)),
 		OpenAPIGetter: o.DiscoveryClient,
 	}
 
-	// include the uninitialized objects by default if --prune is true
-	// unless explicitly set --include-uninitialized=false
-	r := o.Builder.
-		Unstructured().
-		Schema(o.Validator).
-		ContinueOnError().
-		NamespaceParam(o.Namespace).DefaultNamespace().
-		FilenameParam(o.EnforceNamespace, &o.DeleteOptions.FilenameOptions).
-		LabelSelectorParam(o.Selector).
-		Flatten().
-		Do()
-	if err := r.Err(); err != nil {
+	// Initial retrieval of objects to apply before pre-processing.
+	infos, err := o.GetObjects()
+	if err != nil {
 		return err
 	}
+	if len(infos) == 0 {
+		return fmt.Errorf("no objects passed to apply")
+	}
 
-	var err error
-	if o.Prune {
-		o.PruneResources, err = parsePruneResources(o.Mapper, o.PruneWhitelist)
-		if err != nil {
+	if o.preprocessFn != nil {
+		klog.V(4).Infof("Running apply pre-processor function")
+		if err := o.preprocessFn(o); err != nil {
 			return err
 		}
 	}
 
-	output := *o.PrintFlags.OutputFormat
-	shortOutput := output == "name"
-
-	visitedUids := sets.NewString()
-	visitedNamespaces := sets.NewString()
-
-	var objs []runtime.Object
-
-	count := 0
-	err = r.Visit(func(info *resource.Info, err error) error {
-		if err != nil {
-			return err
-		}
+	// The pre-processor might have updated the objects field, so
+	// retrieve these objects again here before applying.
+	infos, _ = o.GetObjects()
```

Comment:
No. I didn't imagine that we would run the resource builder twice. I will add a strong warning comment where the resource builder is currently run that future developers should not run the resource builder twice. The `SetObjects()` method, which is the other method to set the `objects` field, does not clear the `objectsCached` field. And since the `objectsCached` field is not exported, it can't be cleared outside of the package.

Reply:
What kind of changes do we think the preProcessor might do here that require us to the slice of Infos again?

Reply:
Either sorting or filtering the `objects` would force another retrieval before applying.